controller/server/main.go

```diff
@@ -14,25 +14,129 @@
 package main
 
 import (
+	"context"
 	"flag"
 	"fmt"
-	"log"
 	"net"
+	"os"
+	"path/filepath"
 
+	"github.com/google/kne/deploy"
 	cpb "github.com/google/kne/proto/controller"
+	log "github.com/sirupsen/logrus"
+	"k8s.io/client-go/util/homedir"
+
 	"google.golang.org/grpc"
 	"google.golang.org/grpc/credentials/alts"
 )
 
 var (
-	port = flag.Int("port", 50051, "Controller server port")
+	defaultKubeCfg            = ""
+	port                      = flag.Int("port", 50051, "Controller server port")
+	defaultMetallbManifestDir = ""
+	defaultMeshnetManifestDir = ""
 )
 
 type server struct {
 	cpb.UnimplementedTopologyManagerServer
 }
 
+func newDeployment(req *cpb.CreateClusterRequest) (*deploy.Deployment, error) {
+	d := &deploy.Deployment{}
+	switch kind := req.ClusterSpec.(type) {
+	case *cpb.CreateClusterRequest_Kind:
+		d.Cluster = &deploy.KindSpec{
+			Name:                     req.GetKind().Name,
+			Recycle:                  req.GetKind().Recycle,
+			Version:                  req.GetKind().Version,
+			Image:                    req.GetKind().Image,
+			Retain:                   req.GetKind().Retain,
+			GoogleArtifactRegistries: req.GetKind().GoogleArtifactRegistries,
+		}
+	default:
+		return nil, fmt.Errorf("cluster type not supported: %T", kind)
+	}
+	switch metallb := req.IngressSpec.(type) {
+	case *cpb.CreateClusterRequest_Metallb:
+		l := &deploy.MetalLBSpec{}
+		if req.GetMetallb().ManifestDir != "" {
```

Comment:
overwriting default is not a good idea, especially with a long running server, follow this pattern instead:

```
path = defaultMetallbManifestDir
if req.GetMetallb().ManifestDir != "" {
    path = req.GetMetallb().ManifestDir
}
p, err := validatePath(path)
if err != nil {
    return nil, fmt.Errorf("failed to validate manifest path %q", path)
}
l.ManifestDir = p
```

Reply:
done